### ISSUE TITLE
Typescript services scan all models, including disposed, when generating diagnostics

### ISSUE THREAD

**Issue by RoboPhred** [Microsoft/monaco-typescript#10]
After the editor gets a javascript or typescript model, it will parse that model alongside any future models it gets, resulting in confusing errors being indicated.

Below is an example in the [Monaco Playground](https://microsoft.github.io/monaco-editor/playground.html#extending-language-services-configure-javascript-defaults).

The example sets up javascript to use semantic validation, then sets its default model to a constant of type string.  Then, an action is introduced under the name of "Test swap model".  This action will first set the editor to a new model defining a constant of the same name to a number.  Then, it will dispose the old model in an attempt to ensure the editor no longer uses it.

What ends up happening is the old model stays around somehow, and the typescript language services include it when attempting to generate diagnostics.  When it sees the constant defined twice, it complains.

```
// validation settings
monaco.languages.typescript.javascriptDefaults.setDiagnosticsOptions({
	noSemanticValidation: false,
	noSyntaxValidation: false
});

// compiler options
monaco.languages.typescript.javascriptDefaults.setCompilerOptions({
	target: monaco.languages.typescript.ScriptTarget.ES6,
	allowNonTsExtensions: true,
    noLib: true
});

// extra libraries

var jsCode = [
	"const foo = 'asdf';"
].join('\n');

var editor = monaco.editor.create(document.getElementById("container"), {
	value: jsCode,
	language: "javascript"
});


editor.addAction({
    id: "test_swap_model",
    label: "Test swap model",
    run: (editor) => {
        var mod = editor.getModel();
        editor.setModel(monaco.editor.createModel('const foo = 32', 'javascript'));
        mod.dispose();
    }
});
```

**Comment by RoboPhred** [comment]
Simpler way to cause it: Just changing the model twice will do it.

```
// validation settings
monaco.languages.typescript.javascriptDefaults.setDiagnosticsOptions({
    noSemanticValidation: false,
    noSyntaxValidation: false
});

// compiler options
monaco.languages.typescript.javascriptDefaults.setCompilerOptions({
    target: monaco.languages.typescript.ScriptTarget.ES6,
    allowNonTsExtensions: true,
    noLib: true
});

// extra libraries

var jsCode = "const foo = 'asdf';";

var newCode = 'const foo = 32';

var editor = monaco.editor.create(document.getElementById("container"), {
    value: jsCode,
    language: "javascript"
});

editor.setModel(monaco.editor.createModel('const foo = 42;', 'javascript'));
editor.setModel(monaco.editor.createModel('const foo = "foo";', 'javascript'));
```

**Comment by RoboPhred** [comment]
Did some more research into this, I don't think this is the bug I first thought:

First, the second example I gave seems to be intended.  I found the documentation for editor.setModel and it claims that models passed in using it must be disposed on the outside.  If monaco-typescript intentionally parses all models, then this is expected behavior, and disposing the model before setting the new one will handle it:
```
// This works fine
const model = monaco.editor.createModel('const foo = 42;', 'javascript');
editor.setModel(model);
editor.setModel(monaco.editor.createModel('const foo = "foo";', 'javascript'));
model.dispose();
```

However, if the model disposal is delayed, then the error appears
```
const model = monaco.editor.createModel('const foo = 42;', 'javascript');
editor.setModel(model);
editor.setModel(monaco.editor.createModel('const foo = "foo";', 'javascript'));
setTimeout(() => model.dispose(), 1000);
```

Interestingly, causing another diagnostics pass by changing anything in the editor will cause the errors to disappear.  This also happens in my original example.
Because of this, I suspect the real issue is monaco-typescript does not invalidate when a model is disposed.
